packages/@uppy/core/src/BasePlugin.js

```diff
@@ -0,0 +1,98 @@
+/**
+ * Core plugin logic that all plugins share.
+ *
+ * BasePlugin does not contain DOM rendering so it can be used for plugins
+ * without an user interface.
+ *
+ * See `Plugin` for the extended version with Preact rendering for interfaces.
+ *
+ * @param {object} main Uppy core object
+ * @param {object} object with plugin options
+ * @returns {Array|string} files or success/fail message
+ */
+module.exports = class BasePlugin {
+  constructor (uppy, opts) {
+    this.uppy = uppy
+    this.opts = opts || {}
+
+    this.update = this.update.bind(this)
+    this.install = this.install.bind(this)
+    this.uninstall = this.uninstall.bind(this)
+  }
+
+  getPluginState () {
+    const { plugins } = this.uppy.getState()
+    return plugins[this.id] || {}
+  }
+
+  setPluginState (update) {
+    const { plugins } = this.uppy.getState()
+
+    this.uppy.setState({
+      plugins: {
+        ...plugins,
+        [this.id]: {
+          ...plugins[this.id],
+          ...update,
+        },
+      },
+    })
+  }
+
+  setOptions (newOpts) {
+    this.opts = { ...this.opts, ...newOpts }
+    this.setPluginState() // so that UI re-renders with new options
+  }
+
+  update (state) {
+    if (typeof this.el === 'undefined') {
+      return
+    }
+
+    if (this.updateUI) {
+      this.updateUI(state)
+    }
+  }
+
+  unmount () {
+    if (this.isTargetDOMEl && this.el && this.el.parentNode) {
+      this.el.parentNode.removeChild(this.el)
+    }
+  }
```

Comment:
I think these are also DOM related, maybe they belong in UIPlugin? If necessary we could keep the method names in BasePlugin with empty implementations and override them in UIPlugin with this implementation

Reply:
Done, also made `uninstall` empty https://github.com/transloadit/uppy/pull/2944/commits/cbede92ebe2a03d9173d3b799524837d066001cc#diff-5aa90a10b825a1fc796f122d942fa9de85fdfee6e4b0da0958df2a2b07c86938L79, because `unmount` has been moved.